### PR TITLE
Arbitrary Code Execution through third-party package context

### PR DESCRIPTION
**Pull request by BlueSquare1**
Arbitrary Code Execution through third-party package context

### CODE REVIEW

**Review by mohsinenar — CHANGES_REQUESTED** [review]
Missing tests.

**Review by mohsinenar — APPROVED** [review]

File: app/src/test/java/co/ostorlab/insecure_app/BugRuleCallerTest.java [hunk 1]
```diff
@@ -234,4 +235,12 @@ public void ruleCaller_BiometricFingerprintPromptVulnerability_NoExceptionThrown
 
         Assert.assertEquals(caller.getRules().size(), 1);
     }
+
+    @Test
+    public void ruleCaller_PackageContext_NoExceptionThrown() throws Exception{
```
Comment (3asm) [inline app/src/test/java/co/ostorlab/insecure_app/BugRuleCallerTest.java:240]: Please update the markdown to reference this vulnerability and explain it.

File: app/src/main/java/co/ostorlab/insecure_app/bugs/calls/PackageContextCall.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+            if (packageName.startsWith("co.ostorlab.plugins.")) {
+
+                try {
+                    Context packageContext = context.createPackageContext(packageName,
+                            Context.CONTEXT_INCLUDE_CODE | Context.CONTEXT_IGNORE_SECURITY);
+
+                    Class<?> loaderClass = packageContext.getClassLoader().loadClass("co.ostorlab.plugins.camera.Loader");
+                    Method updateMethod = loaderClass.getMethod("Update", Context.class);
+                    updateMethod.invoke(null, packageContext);
```
Comment (3asm) [inline app/src/main/java/co/ostorlab/insecure_app/bugs/calls/PackageContextCall.java]: Shouldn't these be controllable to have a vulnerability?

**Review by 3asm — CHANGES_REQUESTED** [review]

File: app/src/main/java/co/ostorlab/insecure_app/bugs/calls/PackageContextCall.java (hunk 2 above)
Comment (BlueSquare1) [reply app/src/main/java/co/ostorlab/insecure_app/bugs/calls/PackageContextCall.java]: Not necessarily, the attacker can mimic the loaded package information and have the application invoke a method from a rogue package in its context, the attack still however requires the malicious package to be installed on the victim's device, hence the limited impact of this vulnerability.

File: app/src/test/java/co/ostorlab/insecure_app/BugRuleCallerTest.java (hunk 1 above)
Comment (3asm) [reply app/src/test/java/co/ostorlab/insecure_app/BugRuleCallerTest.java:240]: Ping on this comment.

Comment (BlueSquare1) [reply app/src/test/java/co/ostorlab/insecure_app/BugRuleCallerTest.java:240]: https://github.com/Ostorlab/KB/pull/28

Comment (3asm) [reply app/src/test/java/co/ostorlab/insecure_app/BugRuleCallerTest.java:240]: I meant the README of this repo: https://github.com/Ostorlab/ostorlab_insecure_android_app/blob/master/README.markdown

**Review by 3asm — APPROVED** [review]

**Pull request merged**